src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/ValidatorProcessor.cs

```diff
@@ -0,0 +1,236 @@
+ *
+ * D. system software or apparatus including but not limited to losses caused by errors
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+using NpgsqlTypes;
+using RadixDlt.NetworkGateway.Abstractions;
+using RadixDlt.NetworkGateway.Abstractions.Model;
+using RadixDlt.NetworkGateway.Abstractions.Numerics;
+using RadixDlt.NetworkGateway.PostgresIntegration.Models;
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using CoreModel = RadixDlt.CoreApiSdk.Model;
+
+namespace RadixDlt.NetworkGateway.PostgresIntegration.LedgerExtension;
+
+internal record struct ValidatorPublicKeyLookup(long ValidatorEntityId, PublicKeyType PublicKeyType, ValueBytes PublicKey);
+
+internal record ValidatorActiveSet(long Epoch, IDictionary<ValidatorPublicKeyLookup, TokenAmount> ValidatorSet, long StateVersion);
+
+internal class ValidatorProcessor
+{
+    private readonly ProcessorContext _context;
+    private readonly ReferencedEntityDictionary _referencedEntities;
+
+    private List<ValidatorActiveSet> _changes = new();
+
+    private Dictionary<ValidatorPublicKeyLookup, long> _seenPublicKeys = new();
+    private Dictionary<ValidatorPublicKeyLookup, ValidatorPublicKeyHistory> _existingPublicKeys = new();
+
+    private List<ValidatorPublicKeyHistory> _publicKeysToAdd = new();
+    private List<ValidatorActiveSetHistory> _activeSetsToAdd = new();
+
+    public ValidatorProcessor(ProcessorContext context, ReferencedEntityDictionary referencedEntities)
+    {
+        _context = context;
+        _referencedEntities = referencedEntities;
+    }
+
+    public void VisitUpsert(CoreModel.Substate substateData, ReferencedEntity referencedEntity, long stateVersion, long? passingEpoch)
+    {
+        if (substateData is CoreModel.ConsensusManagerRegisteredValidatorsByStakeIndexEntrySubstate entry)
+        {
+            var av = entry.Value.ActiveValidator;
+            var lookup = new ValidatorPublicKeyLookup(_referencedEntities.Get((EntityAddress)av.Address).DatabaseId, av.Key.KeyType.ToModel(), av.Key.GetKeyBytes());
+
+            _seenPublicKeys.TryAdd(lookup, stateVersion);
```

Comment:
why do we scan two different states for public keys? are both of them allowed to mention on ledger that public key for first time ?

Reply:
Clarified.